src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.TypeParameterUsageChecker.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+partial class DelegateCacheRewriter
+{
+    /// <summary>
+    /// Checks if a type or a method is constructed with some type that involves the specified generic type parameters.
+    /// This checker is different with <see cref="TypeSymbolExtensions.ContainsTypeParameters"/> because here we check method symbols, anonmous types, as well as custom modifiers.
+    /// </summary>
+    sealed class TypeParameterUsageChecker : CSharpSymbolVisitor<HashSet<TypeParameterSymbol>, bool>
```

Comment:
>TypeParameterUsageChecker [](http://example.com/codeflow?start=17&length=25)

I do not think we need a special type. The is already a helper ```public static bool ContainsTypeParameters(this TypeSymbol type, HashSet<TypeParameterSymbol> parameters)``` in ```TypeSymbolExtensions```

Reply:
I experimented the helper previously and it was not ideal. It does not check `MethodSymbol`s and does not check everything causing some test failures. See https://github.com/dotnet/roslyn/pull/57882/commits/230471280c1135b12a0ad17449021ada25985354. Either we improve the helper or keep the visitor.

Reply:
> I experimented the helper previously and it was not ideal. It does not check MethodSymbols and does not check everything causing some test failures. See [https://github.com/dotnet/roslyn/pull/57882/commits/230471280c1135b12a0ad17449021ada25985354](https://github.com/dotnet/roslyn/pull/57882/commits/230471280c1135b12a0ad17449021ada25985354). Either we improve the helper or keep the visitor.

>It does not check MethodSymbols

The helper doesn't need to check ```MethodSymbol```. We can call the helper for all the interesting types as it is done in the ```VisitMethod``` method below.

>and does not check everything causing some test failures. See [https://github.com/dotnet/roslyn/pull/57882/commits/230471280c1135b12a0ad17449021ada25985354](https://github.com/dotnet/roslyn/pull/57882/commits/230471280c1135b12a0ad17449021ada25985354).

If you can provide specifics about what it doesn't check, I will be able to recommend something. A test failure on its own gives me pretty much zero information.

> Either we improve the helper or keep the visitor.

It is quite possible that I am missing something, but in this case it is unlikely. I am pretty confident that we don't want to have and maintain yet another symbol visitor. It is likely to require code duplication, initially and in the future. Also, a visitor is susceptible to stack overflows on some deeply nested types and the helper already deals with that the way we would like.

Bottom line, a custom symbol visitor is a non-starter.

Reply:
> If you can provide specifics about what it doesn't check

Method symbols, anonymous types and custom modifiers. Here's one failing case:

```C#
using System;
class C
{
    static void Main(string[] args) => D.Test(0);
}
class D
{
    public static void Test<T>(T t) => Invoke(new { x = t }, Target);

    static void Invoke<T>(T t, Action<T> f) => f(t);

    static void Target<T>(T t) => Console.WriteLine(t);
}
```

For `D.Test<T>`, it calls `Invoke<'a>` where `'a` is `{ T x }`, then creates `Action<'a>` for `Target<'a>`, the helper failed to spot there's a `T` in `'a`.

Reply:
> Method symbols, anonymous types and custom modifiers.

I think I already responded to method symbols above. We don't need it to handle method symbols.

Reply:
> anonymous types

This is definitely a flaw in 
```
        public static TypeSymbol? VisitType<T>(
            this TypeWithAnnotations typeWithAnnotationsOpt,
            TypeSymbol? type,
            Func<TypeWithAnnotations, T, bool, bool>? typeWithAnnotationsPredicate,
            Func<TypeSymbol, T, bool, bool>? typePredicate,
            T arg,
            bool canDigThroughNullable = false,
            bool useDefaultType = false)
```

We need to adjust it to dig into anonymous types. I think this should be done unconditionally.

>and custom modifiers.

Similarly, it feels like we need to add a way for ```VisitType```  to dig into custom modifiers, but do that conditionally. Then, we can add an optional parameter to ```ContainsTypeParameters``` to indicate if we want it to dig into custom modifiers and propagate it to ```VisitType``` accordingly.